src/NLog/Targets/AsyncTaskTarget.cs

```diff
@@ -0,0 +1,378 @@
+// are met:
+// 
+// * Redistributions of source code must retain the above copyright notice, 
+//   this list of conditions and the following disclaimer. 
+// 
+// * Redistributions in binary form must reproduce the above copyright notice,
+//   this list of conditions and the following disclaimer in the documentation
+//   and/or other materials provided with the distribution. 
+// 
+// * Neither the name of Jaroslaw Kowalski nor the names of its 
+//   contributors may be used to endorse or promote products derived from this
+//   software without specific prior written permission. 
+// 
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE 
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE 
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE 
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR 
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS 
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN 
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) 
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF 
+// THE POSSIBILITY OF SUCH DAMAGE.
+// 
+
+namespace NLog.Targets
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Threading;
+
+#if !NET3_5 && !SILVERLIGHT4
+    using System.Threading.Tasks;
+    using NLog.Common;
+    using System.ComponentModel;
+
+    /// <summary>
+    /// Abstract Target with async Task support
+    /// </summary>
+    public abstract class AsyncTaskTarget : Target
+    {
+        private readonly Timer _taskTimeoutTimer;
+        private CancellationTokenSource _cancelTokenSource;
+        private readonly Queue<AsyncLogEventInfo> _requestQueue;
+        private readonly Action _taskStartNext;
+        private readonly Action _taskCancelledToken;
+        private readonly Action<Task, object> _taskCompletion;
+        private Task _previousTask;
+
+        /// <summary>
+        /// How many seconds a Task is allowed to run before it is cancelled.
+        /// </summary>
+        [DefaultValue(150)]
+        public int TaskTimeoutSeconds { get; set; }
+
+        /// <summary>
+        /// Task Scheduler used for processing async Tasks
+        /// </summary>
+        protected virtual TaskScheduler TaskScheduler { get { return TaskScheduler.Default; } }
+
+        /// <summary>
+        /// Constructor
+        /// </summary>
+        protected AsyncTaskTarget()
+        {
+            TaskTimeoutSeconds = 150;
+
+            _taskStartNext = () => TaskStartNext(null);
+            _taskCompletion = TaskCompletion;
+            _taskCancelledToken = TaskCancelledToken;
+            _taskTimeoutTimer = new Timer(TaskTimeout);
```

Comment:
Not sure I understand. TaskTimeout is the callback method, and the timer is started at this line:
`_taskTimeoutTimer.Change(TaskTimeoutSeconds * 1000, Timeout.Infinite);`
Anyway just remove the timer, then and see if the timeout test works :)